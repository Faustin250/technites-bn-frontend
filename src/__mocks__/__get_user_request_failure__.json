{
    "status": 404,
    "error": "This user doesn't have any available requests!"
}