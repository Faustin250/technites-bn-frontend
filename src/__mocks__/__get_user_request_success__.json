{
  "response": {
    "data": [
      {
        "status": 200,
        "message": "user requests",
        "data": [
          {
            "id": 12,
            "user_id": 9,
            "request_type": "OneWay",
            "location_id": 1,
            "departure_date": "2019-11-16 11:41:42.974 +00:00",
            "return_date": "2019-12-20 11:41:42.974 +00:00",
            "destinations": [
              {
                "check_in": "2019-11-09T11:41:42.973Z",
                "check_out": "2019-11-16T11:41:42.973Z",
                "destination_id": 2,
                "accomodation_id": 1
              }
            ],
            "reason": "Visit Uganda",
            "status": "Rejected",
            "createdAt": "2019-11-16T11:41:42.974Z",
            "updatedAt": "2019-11-16T11:41:42.974Z"
          }
        ]
      }
    ]
    
  }
}