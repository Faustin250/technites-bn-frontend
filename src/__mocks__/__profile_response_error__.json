{
  "status": 400,
  "response": {
    "data": {
      "status": 400,
      "error": "User id must be an Integer"
    }
  }
}