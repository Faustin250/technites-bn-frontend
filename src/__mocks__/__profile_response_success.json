{
  "status": 200,
  "response": {
    "data": {
      "status": 200,
      "message": "User found!",
      "data": {
        "id": 5,
        "firstname": "firstname",
        "lastname": "user",
        "email": "requester@request.com",
        "username": "requesting",
        "is_verified": true,
        "role_value": 1,
        "phone": null,
        "gender": "Male",
        "dob": null,
        "address": null,
        "country": null,
        "language": null,
        "currency": null,
        "image_url": "https://avatars0.githubusercontent.com/u/11447549?s=460&v=4",
        "company": null,
        "department": null,
        "line_manager": "manager@admin.com",
        "isEmailAllowed": "true",
        "createdAt": "2019-10-11T10:50:28.214Z",
        "updatedAt": "2019-11-12T09:49:06.152Z"
      }
    }
  }
}