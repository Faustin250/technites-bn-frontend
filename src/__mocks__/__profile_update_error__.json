{
  "status": 400,
  "response": {
    "data": {
      "status": 400,
      "errors": [
        "genders is not allowed"
      ]
    }
  }
}