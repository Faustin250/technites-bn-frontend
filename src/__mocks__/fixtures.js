export const loginPayload = {
  token: "test token",
  user: { email: "test user" },
};

export const payload = {
  status: 20,
  message: "Sign up the",
  data: {
    id: 28,
    email: "amilylisy@yahoo.com",
    firstname: "amily",
    lastname: "kassim",
    role_value: 1,
    username: "amilykassim",
  },
  token: "eyJhbGcibK-l8Ic4",
};

export const token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTAsImZpcnN0bmFtZSI6IkZyZWQiLCJsYXN0bmFtZSI6Ik11Y3lvIiwiZW1haWwiOiJtdWN5b21pbGxlckBnbWFpbC5jb20iLCJ1c2VybmFtZSI6bnVsbCwiaXNfdmVyaWZpZWQiOnRydWUsInJvbGVfdmFsdWUiOjEsInBob25lIjpudWxsLCJnZW5kZXIiOiJNYWxlIiwiZG9iIjoiMTk5NC0wNS0wMlQwMDowMDowMC4wMDBaIiwiYWRkcmVzcyI6Im51bGwiLCJjb3VudHJ5IjoibnVsbCIsImxhbmd1YWdlIjoibnVsbCIsImN1cnJlbmN5IjoibnVsbCIsImltYWdlX3VybCI6Imh0dHA6Ly9yZXMuY2xvdWRpbmFyeS5jb20vdGVjaG5pdGVzL2ltYWdlL3VwbG9hZC92MTU3MzgwODYzNy9tdHd3a2ExZHh0aWVrd3Nncmlvai5qcGciLCJjb21wYW55IjoiQW5kZWxhIiwiZGVwYXJ0bWVudCI6IkVuZ2luZWVyaW5nIiwibGluZV9tYW5hZ2VyIjoicnVndW1iaXJham9yZHliYXN0aWVuQGdtYWlsLmNvbSIsImlzRW1haWxBbGxvd2VkIjoidHJ1ZSIsImlhdCI6MTU3MzgwODg1NiwiZXhwIjoxNTczODk1MjU2fQ._fvFcTCo7YdJ1yL_X_iVFOrby5k5b3bcrOb0f9DQpEo";

export const request = {
  User: {
    id: 5,
    firstname: "Dennis",
    lastname: "Lohan",
    username: "requesting",
    email: "requester@request.com",
  },
  id: 9,
  user_id: 5,
  request_type: "ReturnTrip",
  location_id: 1,
  departure_date: "2020-09-25",
  status: "Pending",
};

export const requests = [request];
