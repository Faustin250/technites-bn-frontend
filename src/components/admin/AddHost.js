/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/no-unused-state */
/* eslint-disable react/prop-types */
/* eslint-disable camelcase */
/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { Component } from "react";
import { connect } from "react-redux";
import { toast } from "react-toastify";
import Joi from "joi-browser";
import _ from "lodash";
import { retrieveProfile } from "../../redux/actions/profileAction";
import { addHost } from "../../redux/actions/AdminActions";
import HomeNav from "../home-nav/HomeNav";
import SideBar from "../side-bar";
import Footer from "../footer";
import "./AddHost.scss";
import Input from "../profile-page/Input";
import Button from "../common/Button";


export class AddHost extends Component {
    schema = {
      email: Joi.string()
        .min(4)
        .email({ minDomainSegments: 2 })
        .label("Email")
        .required(),
      firstname: Joi.string()
        .min(3)
        .label("Firstname")
        .required(),
      lastname: Joi.string()
        .min(3)
        .label("Lastname")
        .required(),
    };

    constructor(props) {
      super(props);
      this.state = {
        email: "",
        firstname: "",
        lastname: "",
        errors: {},
        user: {},
        hostAdded: false,
      };
      this.onChange = this.onChange.bind(this);
      this.onSubmit = this.onSubmit.bind(this);
    }

    componentDidMount() {
      this.props.retrieveProfile();
    }

    UNSAFE_componentWillReceiveProps({ errors, user, admin }) {
      if (user.role_value < 7) {
        window.location.href = "/dashboard";
      }
      if (admin.hostAdded) {
        toast.success("Host Added succesfully", {
          position: toast.POSITION.TOP_RIGHT,
        });
        window.location.href = "/addHost";
      }
      this.setState({ errors, user });
      if (Array.isArray(errors.errors)) {
        toast.error("Please fill in all the fields", {
          position: toast.POSITION.TOP_RIGHT,
        });
      } else {
        toast.error(errors.message ? errors.message : errors.error, {
          position: toast.POSITION.TOP_RIGHT,
        });
      }
    }

    onChange(event) {
      const { target } = event;
      const { errors } = this.state;
      const { value } = target;
      const { name } = target;
      const err = this.validateHost(name, value);
      if (err) errors[name] = err;
      else delete errors[name];
      this.setState({
        [name]: value,
        errors,
      });
    }

    onSubmit(e) {
      e.preventDefault();
      const host = {
        email: this.state.email,
        firstname: this.state.firstname,
        lastname: this.state.lastname,
      };
      this.props.addHost(host);
    }

    validateHost = (name, value) => {
      const obj = { [name]: value };
      const schema = { [name]: this.schema[name] };
      const { error } = Joi.validate(obj, schema);
      return error ? error.details[0].message : null;
    };

    render() {
      const { user } = this.props;
      const { errors } = this.state;
      return (
        <>
          <HomeNav user={user} />
          <SideBar userRole={user.role_value} />
          <div className="page-info">
            <h1 className="page-title">Add Host</h1>
            <h4 className="sub-title">
              {" "}
              <span className="sub-title-info">
                <a href="#">Dashboard </a>
              </span>
            /
              <span className="sub-title-info">
                <a href="#"> Add Host</a>
              </span>
            </h4>
          </div>
          <div className="wrapper">
            <form noValidate onSubmit={this.onSubmit}>
              <Input
                id="email"
                mode
                type="text"
                name="email"
                label="Email"
                handler={this.onChange}
                error={errors.email}
              />
              <Input
                id="firstname"
                mode
                type="text"
                name="firstname"
                label="Firstname"
                handler={this.onChange}
                error={errors.firstname}
              />
              <Input
                id="lastname"
                mode
                type="text"
                name="lastname"
                label="Lastname"
                handler={this.onChange}
                error={errors.lastname}
              />
              <Button type="submit" label="Add Host" id="add-host" />
            </form>
          </div>
          <Footer />
        </>
      );
    }
}
const mapStateToProps = (state) => ({
  errors: state.errors,
  user: state.profile.user,
  admin: state.admin,
});
export default connect(mapStateToProps, { retrieveProfile, addHost })(AddHost);
