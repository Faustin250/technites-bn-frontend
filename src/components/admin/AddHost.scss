@import "../../styles/variables";
.wrapper {
  position: absolute;
  background-color: #fff;
  padding: 50px 20px 50px 20px;
  width: 80%;
  height: auto;
  margin: 0px 20px 0px 10px;
  &--addhost {
    width: 78vw !important;
    top: 10rem;
    right: -1rem;
  }
}
.page-info {
    margin: 20px 0px 20px 50px;
    .page-title {
      color: $textColor;
    }
    .sub-title {
      .sub-title-info {
        a {
          text-decoration: none;
          color: $primaryColor;
        }
      }
    }
  }