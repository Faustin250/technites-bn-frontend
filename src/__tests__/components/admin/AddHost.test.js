/* eslint-disable import/no-named-as-default */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable no-unused-vars */
/* eslint-disable no-multi-assign */
/* eslint-disable no-undef */
import React from "react";
import Enzyme, { mount, shallow } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import Adapter from "enzyme-adapter-react-16/build";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import toJson from 'enzyme-to-json';
import AddHost, {
  AddHost as AddHostComponent,
} from "../../../components/admin/AddHost";
import successresponse from "../../../__mocks__/__get_user_request_success__.json";
import { messages } from "../../../__mocks__/fixtures";
const validateHost = jest.fn();

Enzyme.configure({ adapter: new Adapter() });
const mockedStore = configureStore([thunk]);
const addHost = jest.fn();
const props = {
  messages,
  notifications: {
    notifications: [
      {
        createdAt: "2019-11-13T10:07:21.401Z",
        id: 8,
        message: "visit nairobi",
        request_id: 12,
        seen: "false",
        type: "ReturnTrip",
        updatedAt: "2019-11-15T13:13:44.347Z",
        user_id: 49,
        notPaneDisplay: false,
      },
    ],
    markThisRead: jest.fn(),
    toggleNotDisplay: jest.fn(),
    notPaneDisplay: false,
    notificationCount: 1,
  },
  loginState: {
    isAuthenticated: true,
  },
  profile: {
    user: {
      firstname: "Rugumbira",
      image_url:
      "https://res.cloudinary.com/dodfpnbik/image/upload/v1574070442/Screen_Shot_2019-11-18_at_11.44.38_bdjv7r.png",
      role_value: 7,
    },
  },
  user: {
    firstname: "Rugumbira",
    image_url:
    "https://res.cloudinary.com/dodfpnbik/image/upload/v1574070442/Screen_Shot_2019-11-18_at_11.44.38_bdjv7r.png",
    role_value: 7,
  },
  postsPerPage: 4,
  currentPage: 1,
  errors: {},
  addHost,
  retrieveProfile: jest.fn(),
  requests: {
    requestFound: true,
    requests: [
      {
        id: 8,
        user_id: 9,
        request_type: "OneWay",
        location_id: 3,
        departure_date: "2019-11-16",
        return_date: null,
        destinations: [
          {
            room_id: 2,
            check_in: "2019-11-18T15:16:38.447Z",
            check_out: "2019-12-27T15:16:38.447Z",
            destination_id: 4,
            accomodation_id: 1,
          },
        ],
        reason: "Visit Lagos",
        status: "Pending",
        createdAt: "2019-11-11T06:53:54.602Z",
        updatedAt: "2019-11-11T06:53:54.602Z",
        User: {
          image_url: "image",
          email: "usertest@gmail.com",
          phone: "123456",
          gender: "male",
          address: "address",
          country: "country",
          language: "language",
          company: "company",
          department: "department",
        },
      },
      {
        id: 8,
        user_id: 9,
        request_type: "OneWay",
        location_id: 3,
        departure_date: "2019-11-16",
        return_date: null,
        destinations: [
          {
            room_id: 2,
            check_in: "2019-11-18T15:16:38.447Z",
            check_out: "2019-12-27T15:16:38.447Z",
            destination_id: 4,
            accomodation_id: 1,
          },
        ],
        reason: "Visit Lagos",
        status: "Accepted",
        createdAt: "2019-11-11T06:53:54.602Z",
        updatedAt: "2019-11-11T06:53:54.602Z",
        User: {
          image_url: "image",
          email: "usertest@gmail.com",
          phone: "123456",
          gender: "male",
          address: "address",
          country: "country",
          language: "language",
          company: "company",
          department: "department",
        },
      },
    ],
  },
};

const store = mockedStore(props);

describe("Admin Panel for Adding Hosts", () => {
  const component = mount(
    <Provider store={store}>
      <Router>
        <AddHostComponent {...props} history={{ push: jest.fn() }} />
      </Router>
        ,
    </Provider>,
  );
  const component2 = shallow(
    <AddHostComponent {...props} history={{ push: jest.fn() }} />,
  );
  it("should render without crashing", () => {
    const nextPropsSuccess = { ...props };
    component.setProps(nextPropsSuccess);
    expect(component).toHaveLength(1);
  });
  it("should render without crashing || snapshot", () => {
    expect(toJson(component2)).toMatchSnapshot();
  });
  it("should trigger componentWillReceiveProps", () => {
    const nextPropsError = {
      errors: {
        message: "Error message",
      },
      user: {
        role_value: 1,
      },
      admin: {
        hostAdded: true,
      },
    };
    component2.setProps(nextPropsError);
    expect(component.props()).toHaveProperty("admin", props.admin);
  });
  it("should trigger componentWillReceiveProps when user is superAdmin", () => {
    const nextPropsError = {
      errors: {
        message: "Error message",
      },
      user: {
        role_value: 7,
      },
      admin: {
        hostAdded: true,
      },
    };
    component2.setProps(nextPropsError);
    expect(component.props()).toHaveProperty("user", props.user);
  });
  it("should trigger componentWillReceiveProps when host is not added", () => {
    const nextPropsError = {
      errors: {
        message: "Error message",
      },
      user: {
        role_value: 7,
      },
      admin: {
        hostAdded: false,
      },
    };
    component2.setProps(nextPropsError);
    expect(component.props()).toHaveProperty("admin", props.admin);
  });
  it("should trigger componentWillReceiveProps when there is an array of errors", () => {
    const nextPropsError = {
      errors: {
        errors: [
          { message: "error message" },
          { message: "error message" },
        ],
      },
      user: {
        role_value: 7,
      },
      admin: {
        hostAdded: false,
      },
    };
    component2.setProps(nextPropsError);
    expect(component.props()).toHaveProperty("user", props.user);
  });
  it("should trigger componentWillReceiveProps when there is an error instead of a message", () => {
    const nextPropsError = {
      errors: {
        error: "Error message",
      },
      user: {
        role_value: 7,
      },
      admin: {
        hostAdded: true,
      },
    };
    component2.setProps(nextPropsError);
    expect(component.props()).toHaveProperty("errors", props.errors);
  });
  it("should simulate handleInputChange", () => {
    component2.find("#firstname").props().handler({
      target: { name: "firstname", value: "test" },
    });
    expect(component2.state().firstname).toBe(
      "test",
    );
  });
  it("should simulate handleInputChange catch an error", () => {
    component2.find("#email").props().handler({
      target: { name: "email", value: "test" },
    });
    expect(component2.state().email).toBe(
      "test",
    );
  });
  it("should simulate click", () => {
    component.find("form").simulate("submit");
    expect(addHost).toHaveBeenCalled();
  });
});
