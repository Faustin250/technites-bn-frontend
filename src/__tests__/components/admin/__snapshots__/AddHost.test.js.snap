// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Panel for Adding Hosts should render without crashing || snapshot 1`] = `
<Fragment>
  <Connect(HomeNav) />
  <Connect(SideBar) />
  <div
    className="main-page"
  >
    <div
      className="page-info"
    >
      <h1
        className="page-title"
      >
        Add Host
      </h1>
      <h4
        className="sub-title"
      >
         
        <span
          className="sub-title-info"
        >
          <a
            href="#"
          >
            Dashboard 
          </a>
        </span>
        /
        <span
          className="sub-title-info"
        >
          <a
            href="#"
          >
             Add Host
          </a>
        </span>
      </h4>
    </div>
    <div
      className="wrapper wrapper--addhost"
    >
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <Input
          error=""
          handler={[Function]}
          id="email"
          label="Email"
          mode={true}
          name="email"
          type="text"
          value=""
        />
        <Input
          error=""
          handler={[Function]}
          id="firstname"
          label="Firstname"
          mode={true}
          name="firstname"
          type="text"
          value=""
        />
        <Input
          error=""
          handler={[Function]}
          id="lastname"
          label="Lastname"
          mode={true}
          name="lastname"
          type="text"
          value=""
        />
        <Button
          id="add-host"
          label="Add Host"
          type="submit"
        />
      </form>
    </div>
  </div>
  <Footer />
</Fragment>
`;
